### PR TITLE
[R] [CI] use lintr 3.1.0

### PR DESCRIPTION
**Pull request by jameslamb**
`{lintr}` 3.1.0 came out a few weeks ago: https://github.com/r-lib/lintr/releases/tag/v3.1.0

It includes a few new linters, a bunch of bugfixes, and some new deprecation warnings. `{xgboost}` has already been using it, since this instalation is unpinned:

https://github.com/dmlc/xgboost/blob/a57371ef7c58acefe4bc794a23089a90734cc1a1/R-package/tests/helper_scripts/install_deps.R#L19

This PR proposes some changes to the R linting to better take advantage of that new version.

Fixes these warnings

```text
1: Linter no_tab_linter was deprecated in lintr version 3.1.0. Use whitespace_linter instead.
2: Linter unneeded_concatenation_linter was deprecated in lintr version 3.1.0. Use unnecessary_concatenation_linter instead.
```

And these errors

```text
[[1]]
R-package/demo/interaction_constraints.R:47:25: warning: [lengths] Use lengths() to find the length of each element in a list.
  interaction_length <- sapply(interaction_list, length)
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

[[2]]
R-package/R/xgb.model.dt.tree.R:90:7: warning: [boolean_arithmetic] Use any() to express logical aggregations. For example, replace length(which(x == y)) == 0 with !any(x == y).
      sum(grepl('leaf=(\\d+)', text)) < 1) {
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

[[3]]
R-package/R/xgb.plot.deepness.R:139:24: warning: [lengths] Use lengths() to find the length of each element in a list.
    data.table(Depth = sapply(paths, length), ID = To[Leaf == TRUE])
                       ^~~~~~~~~~~~~~~~~~~~~

[[4]]
R-package/tests/testthat/test_helpers.R:192:18: warning: [matrix] Use rowSums(shapi) rather than apply(shapi, 1, sum)
    expect_equal(apply(shapi, 1, sum), pred, tol = tol)
                 ^~~~~~~~~~~~~~~~~~~~
```

### Notes for Reviewers

We made similar changes a few weeks ago in LightGBM and it's been working well so far: https://github.com/microsoft/LightGBM/pull/5997

Thanks for your time and consideration!

### CODE REVIEW

**Review by trivialfis — APPROVED** [review]
Nice upgrade! Love watching the code being cleaned up.

**Comment by jameslamb** [comment]
Sure, happy to help!
